bpf/bpf_lxc.c

```diff
@@ -1022,6 +1022,16 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 	 */
 	if (is_defined(ENABLE_ROUTING) || hairpin_flow ||
 	    is_defined(ENABLE_HOST_ROUTING)) {
+		/* Hairpin requests need to pass through the backend's to-container
+		 * path, to create a CT_INGRESS entry with .lb_loopback set. This
+		 * drives RevNAT in the backend's from-container path.
+		 *
+		 * Hairpin replies are fully RevNATed in the backend's from-container
+		 * path. Thus they don't match the CT_EGRESS entry, and we can't rely
+		 * on a CT_REPLY result that would provide bypass of ingress policy.
+		 * Thus manually skip the ingress policy path.
+		 */
+		bool bypass_ingress_policy = hairpin_flow && ct_status == CT_REPLY;
```

Comment:
This means we defer bypassing policy enforcement for requests only (TCP syns), replies bypass policy enforcement straightaway. 

Thinking over this a bit more...
I wonder when packets are not passed through stack (e.g., endpoint routes), can we store some context in the packet scratch buffer.

Reply:
>I wonder when packets are not passed through stack (e.g., endpoint routes), can we store some context in the packet scratch buffer.

As optimization to avoid the `ct_has_loopback_egress_entry4()` call you mean? Sure, that might work. Not something I would want to backport though.

I also believe we're still at a stage where we need to streamline the code paths, and produce something that behaves (close to) identically with/without per-EP routes. Fine-tuning can come afterwards.

Reply:
Well, that too. But, mainly so that we could have a consistent behavior for all packets in a flow.

Reply:
Ah yes, agreed. I'm hopeful that we can also sort out the bypass for CT_REPLY during the 1.15 cycle :crossed_fingers:.